test/gtest/ucp/test_ucp_tag.cc

```diff
@@ -462,28 +463,39 @@ UCS_TEST_P(test_ucp_tag_limits, check_max_short_zcopy_thresh_zero, "ZCOPY_THRESH
 UCP_INSTANTIATE_TEST_CASE(test_ucp_tag_limits)
 
 
-class test_ucp_tag_fallback : public ucp_test {
+class test_ucp_tag_nbx : public test_ucp_tag {
 public:
     void init() {
         /* forbid zcopy access because it will always fail due to read-only
          * memory pages (will fail to register memory) */
         modify_config("ZCOPY_THRESH", "inf");
-        ucp_test::init();
-        sender().connect(&receiver(), get_ep_params());
-        receiver().connect(&sender(), get_ep_params());
-    }
-
-    static void get_test_variants(std::vector<ucp_test_variant>& variants) {
-        add_variant(variants, UCP_FEATURE_TAG);
+        test_ucp_tag::init();
+        m_completed = 0;
     }
 
 protected:
     static const size_t MSG_SIZE;
+    uint32_t m_completed;
+
+    static void send_callback(void *req, ucs_status_t status,
+                              void *user_data)
+    {
+        request_free((request*)req);
```

Comment:
can add wrapper for send and recv callback's bodies?

also, what about checking the status of UCP request:
```
EXPECT_UCS_OK(ucp_request_status_check(req));
```

Reply:
will not work: request formally may be in state "IN_PROGRESS"

Reply:
`ucp_request_complete()` guarantees that UCP request's status will be set
but I guess we should set the status in `ucp_request_imm_cmpl_param()` - otherwise user can't rely on the status of the UCP request

@brminich wdyt?

Reply:
we have status as a callback param, why need to check req?

Reply:
> we have status as a callback param, why need to check req?

to align with the non-`NO_IMM_CMPL` behavior?
why then we need to set the status from this case?
also, I think it is possible that the user could wait for the request completion in the following way:
```
void send_cb(void *req, ucs_status_t status, void *user_data)
{
    free(req);
}
...
params.cb.send      = send_cb;
params.request      = ucs_malloc(...);
params.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
                      UCP_OP_ATTR_FLAG_NO_IMM_CMPL |
                      UCP_OP_ATTR_FIELD_REQUEST;
ucs_status_ptr_t send_req = ucp_tag_send_nbx(ep, buf, 1024, 0, &param);
do {
    ucp_worker_progress(worker);
} while (ucp_request_check_status(params.request) == UCS_INPROGRESS);
```
this test
- hang, if send completed immediately inside `ucp_tag_send_nbx()` - `ucp_request_imm_cmpl_param()` completes this request that doenst' set status
- pass, if send didn't complete inside `ucp_tag_send_nbx()` - `ucp_request_complete()` completes this request that sets status
I guess we have to align the behavior for both cases. Since users may rely on the status of the operations by using different ways even if they specified the callback.

Reply:
your test is incorrect: after ```send_cb``` is called call to ```ucp_request_check_status(params.request)``` is not valid

Reply:
> your test is incorrect: after `send_cb` is called call to `ucp_request_check_status(params.request)` is not valid

yes, sorry
```
void send_cb(void *req, ucs_status_t status, void *user_data)
{
    /* yawn */
}
```
and free the memory for request after check the status

Reply:
I do not insist and approved the PR, but just want to check that we probably want to align providing a completion status for different cases

Reply:
In case if request is not deleted in callback, then it will be updated to completed state. User may select when request should be deleted.